### DISCUSSION TITLE
I find myself needing to modify types frequently

### DISCUSSION THREAD

**Discussion by ghost** [prisma/prisma#25237] [Q&A]
### Question

Hey!
First time using Prisma and while I'm generally having quite a good time I find myself having to create types that simply modify the behavior of the type automatically generated by Prisma just to continue having (somewhat) decent type safety.

For this example I want to take my User model which currently looks like this. 

```prisma
model User {
  PKUser               Int    @id(map: "user_pkey") @default(autoincrement())
  Username             String @db.VarChar(16)
  Password             String
  Email                String @db.VarChar(255)
  Role                 Int    @default(1)
  Role_user_RoleToRole Role   @relation("user_RoleToRole", fields: [Role], references: [PKRole], onUpdate: NoAction, map: "user_Role_fkey")
}
```

When working with what `npx prisma generate` generates in the clients for this I have found multiple things that have warranted me to create derived types from this.

1)  I'm globally omitting the Password field from User as to not accidentally leak this to sources outside the server/API itself. This has prompted an issue where specifying the return type of my query functions as User | null was not possible as the return value from queries now doesn't satisfy the shape of User. Thus I created the type SafeUser which is simply a User that omits the field Password

**types.d.ts**
```typescript
type SafeUser = Omit<User, 'Password'>;
```

**user.repository.ts**
```typescript
import { SafeUser } from "../../prisma/types.js";

public async findByPrimaryKey(pk: number): Promise<SafeUser | null>
{
    return this.prisma.user.findFirst({
        where: {
            PKUser: {
                equals: pk
            }
        }
    });
}
```
2) The next problem came when I tried to implement a method to create a new User. The model has two fields which might possibly be undefined with PKUser and Role; the first obviously not being set manually since it is defined as an auto increment and Role having a default value. Again I found myself unable to simply use the generated User type and found myself crafting a type that instead makes these fields optional.

**types.d.ts**
```typescript
type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;
type CreatableUser = Optional<User, 'PKUser' | 'Role'>
```

Which I can then use to type the object I feed into `userRepository.create()`.

I feel like this isn't really a sustainable way to do things going forward, especially creating derived types where certaing things are now optional - then again I did not find a way to do so without making the field optional in the database as well which,  at least in this case, is simply not a solution. 

I'm almost certain I'm overlooking something but none of my own research seemed to be able to help me tackle this problem so I'd really love some input here!

- OS: Windows 11
- Database: Postgres
- Node.js version: 22.5.0

- prisma v5.19.1

**Comment by nurul3101** [comment]
Hello @ghost 👋 

Apologies for the delay in responding to this Discussion, this one slipped through the cracks.

**Regarding Omitting the password field**:

As of Prisma ORM version 5.16.0, there's a built-in way to do this using the [omitApi](https://www.prisma.io/docs/orm/prisma-client/queries/excluding-fields#excluding-a-field-globally-using-omit) Preview feature. Instead of creating a `SafeUser` type, you can use Prisma's global omit feature:
```ts
const prisma = new PrismaClient({
  omit: {
    user: {
      password: true
    }
  }
})
```

This will exclude the password field in all queries, making it safer and eliminating the need for a separate `SafeUser` type.

**Regarding creating a user with optional fields**:

Prisma actually provides a type that makes the auto-incrementing ID and fields with default values optional. Instead of creating your own `CreatableUser` type, you can use `Prisma.UserCreateInput.` This type is automatically generated by Prisma and includes all the necessary optionality.

Something like this:
```ts
import { Prisma } from '@prisma/client'

async function createUser(user: Prisma.UserCreateInput) {
  return prisma.user.create({ data: user })
}
```

I hope this helps! Let me know in case you have any queries.
